pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -90,6 +92,139 @@ pub struct InMemoryLayerInner {
     resource_units: GlobalResourceUnits,
 }
 
+/// Support the same max blob length as blob_io, because ultimately
+/// all the InMemoryLayer contents end up being written into a delta layer,
+/// using the [`crate::tenant::blob_io`].
+const MAX_SUPPORTED_BLOB_LEN: usize = crate::tenant::blob_io::MAX_SUPPORTED_BLOB_LEN;
+const MAX_SUPPORTED_BLOB_LEN_BITS: usize = {
+    let trailing_ones = MAX_SUPPORTED_BLOB_LEN.trailing_ones() as usize;
+    let leading_zeroes = MAX_SUPPORTED_BLOB_LEN.leading_zeros() as usize;
+    assert!(trailing_ones + leading_zeroes == std::mem::size_of::<usize>() * 8);
+    trailing_ones
+};
+
+/// See [`InMemoryLayerInner::index`].
+///
+/// For space-efficiency, this value is a bitfield.
+///
+/// Layout:
+/// - 1 bit: `will_init`
+/// - [`MAX_SUPPORTED_BLOB_LEN_BITS`]: `len`
+/// - [`MAX_SUPPORTED_POS_BITS`]: `pos`
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct InMemoryLayerIndexValue(u64);
+
+impl InMemoryLayerIndexValue {
+    /// Derive remaining space for pos.
+    /// TODO: define and enforce a hard limit at the [`crate::tenant::Timeline::should_roll`] level.
```

Comment:
Is it safe to merge this before we have that hard limit?

Reply:
This comment is stale.

should_roll actually enforces a hard limit `checkpoint_distance`.

Further, this PR adds `validate_checkpoint_distance`, which is called on PS startup and tenant attach.
It's not straight-forward to add it to the config update code path, but, we don't change checkpoint_distance in prod anyways.

Reply:
558442610d8e5d128b20d0ec6c8de442db135088

Reply:
:+1: